lib/topology/platforms/shell.py

```diff
@@ -417,74 +428,70 @@ def get_response(self, connection=None):
 
         return '\n'.join(lines)
 
-    @_check_existing_connection
     def is_connected(self, connection=None):
         """
         See :meth:`BaseShell.is_connected` for more information.
         """
-        if connection is None:
-            return (
-                self._default_connection is not None
-            ) and self._connections[self._default_connection].isalive()
-        else:
-            return self._connections[connection].isalive()
+        # Get connection
+        spawn = self._get_connection(connection)
+        return spawn.isalive()
 
     def connect(self, connection=None):
         """
         See :meth:`BaseShell.connect` for more information.
         """
-        try:
-            if self.is_connected(connection=connection):
-                raise Exception('Shell already connected.')
-        except _NonExistingConnectionError:
-            pass
+        connection = connection or self._default_connection or '0'
 
-        # Evaluate connection value
-        if self._default_connection is None:
-            if connection is None:
-                connection = '0'
-            self._default_connection = connection
+        if connection in self._connections:
```

Comment:
This is wrong. A shell may be in ``self._connections`` but still be disconnected. Previously this was:

    def is_connected(self):
        """
        See :meth:`BaseShell.is_connected` for more information.
        """
        return self._spawn is not None and self._spawn.isalive()

and

    def connect(self):
        """
        See :meth:`BaseShell.connect` for more information.
        """
        if self.is_connected():
            raise Exception('Shell already connected.')

        # Create a child process
        self._spawn = spawn(
            self._get_connect_command().strip(),
            **self._spawn_args
        )